Docs release 1.10.0 csi 4365 host define (#517)
IBM/ibm-block-csi-driver: docs/content/installation/install_hostdefinition.md

# Installing the `HostDefiner` custom resource

> Install the `HostDefiner` custom resource to enable dynamic host definitions from within the CSI driver.

Author: updated in source files - waiting for full installation information

> Install the `HostDefiner` custom resource to enable dynamic host definitions from within the CSI driver.

Author: ready

> For more information on using dynamic host connectivity, see [Using dynamic host connectivity](../using/using_hostdefinition.md).

Reviewer: ```suggestion
For more information on using dynamic host definition, see [Using dynamic host definition](../using/using_hostdefinition.md).
```

Author: updated

Use the following steps to install the `HostDefiner` custom resource, with [GitHub](https://github.com/IBM/ibm-block-csi-operator).

**Note:** Only hosts added after the `HostDefiner` installation are dynamically defined within the CSI driver.

Download the `HostDefiner` customer resource manifest from [GitHub](https://github.com/IBM/ibm-block-csi-operator).

    curl https://raw.githubusercontent.com/IBM/ibm-block-csi-operator/v1.10.0/config/samples/csi_v1_hostdefiner.yaml > csi_v1_hostdefiner.yaml

> After the `HostDefiner` is installed it searches for host on storage definitions that have not yet been defined.

Reviewer: ```suggestion
After the `HostDefiner` is installed it searches for nodes that have not yet been defined on storage, and defines them.
```

> After the `HostDefiner` is installed, it searches for nodes that have not yet been defined on storage, and defines them.

Reviewer: @matancarmeli7 this implies there is no need to redeploy the csi node in order to dynamically create definitions for existing nodes. is this so?

Reviewer 2: that is right

Reviewer: then is the following note true?
>**Note:** Only hosts added after the host definer installation are dynamically defined within the CSI driver.

Reviewer 2: this is not true, the host definer doesn't care when the CSI nodes deployed

Author: i'll delete that note

# Installing the `HostDefinition` custom resource

Install the `HostDefinition` custom resource to enable automatic host definitions from within the CSI driver.

Use the following steps to install the `HostDefinition` custom resource, with [GitHub](https://github.com/IBM/ibm-block-csi-operator).

**Note:** Only hosts added after installation are automatically defined with the CSI driver.

1. Download the manifest from GitHub.

        ```
        curl https://raw.githubusercontent.com/IBM/ibm-block-csi-operator/v1.10.0/deploy/installer/generated/host-definition.yaml > host-definition.yaml
        ```

2. Verify that the hostdefinition is in the _Ready_ phase.

    ```
    $> kubectl get hostdefinition
    NAME                  AGE    PHASE   STORAGE          HOST
    <k8s_resource_name1>  102m   Ready   <storage_name>   <host_name_w1>
    k8s_resource_name2>  102m   Ready   <storage_name>   <host_name_w2>
    ```

    If in an `Error` state, a retry can be forced using TBD.
    
3.  (Optional) Update the prefix and connectivity fields in the `HostDefiner.yaml`.

    |Field|Description|
    |---------|--------|
    |prefix|Adds a prefix to the hosts defined by the CSI driver.|

>     |connectivity|Selects the connectivity type for the host ports.<br>Possible input values are:<br>- `iscsi` for use with iSCSI connectivity<br>- `fc` for use with Fibre Channel over SCSI connectivity<br>- `nvme` for use with NVME over Fibre Channelconnectivity (Spectrum Virtualize storage systems only)<br>By default, this field is blank and the driver selects the strongest of available connectivity options.|

Reviewer: the outcome of the default it is not clear by `strongest`

Author: dominant?

Reviewer: in the market?
I suggest something like the following:
```
By default, this field is blank and the driver selects the first of available connectivity types, according to the following order/hierarchy:
NVMe, FC, iSCSI.
```

Author: updated

Reviewer: `available on the storage system` @matancarmeli7 do we check the availability on storage? don't we check the availability on the worker?

Reviewer 2: we check it on the worker now

Author: updated

# Installing the `HostDefinition` custom resource

> Install the `HostDefinition` custom resource to enable automatic host definitions from within the CSI driver.

Reviewer: isn't it "host defin**er** cr? I thought the host definition is a separate CR

Reviewer 2: you are right

Reviewer: then please sync with @rivkap about this

Author: updated

Reviewer: I still see:
>Installing the `HostDefinition` custom resource

in this PR diff

Author: updated - waiting to check in my changes until more information is received

> Use the following steps to install the `HostDefinition` custom resource, with [GitHub](https://github.com/IBM/ibm-block-csi-operator).

Author: The existing installation sections are not all github based and are specifically for installing the operator and driver.
as far as I understand (@matancarmeli7  please confirm) this can be installed optionally after any installation type.
In any case, as this is a separate installation I believe that it should be documented separately but under the Installation chapter.

> Use the following steps to install the `HostDefinition` custom resource, with [GitHub](https://github.com/IBM/ibm-block-csi-operator).

Reviewer 2: @rivkap yes we can download it in any installation that we want

> Use the following steps to install the `HostDefinition` custom resource, with [GitHub](https://github.com/IBM/ibm-block-csi-operator).

Author: the from a documentation point of view we keep it separate.

> Use the following steps to install the `HostDefinition` custom resource, with [GitHub](https://github.com/IBM/ibm-block-csi-operator).

Reviewer: why?
and either way, we should document per each type of installation

> Use the following steps to install the `HostDefinition` custom resource, with [GitHub](https://github.com/IBM/ibm-block-csi-operator).

Author: is there a different way to install the host definer?

> Use the following steps to install the `HostDefinition` custom resource, with [GitHub](https://github.com/IBM/ibm-block-csi-operator).

Reviewer: I'll let @matancarmeli7 fill you in with the exact details (I haven't done myself)

> Use the following steps to install the `HostDefinition` custom resource, with [GitHub](https://github.com/IBM/ibm-block-csi-operator).

Author: updated per meeting with @roysahar

> **Note:** Only hosts added after installation are automatically defined with the CSI driver.

Reviewer: IMO it's not clear that `hosts added` means added to the csi node pod deployment

Author: changed

> 1.  Install the operator.

Reviewer: which operator?

Author: this was a setup error. fixed.

    1. Download the manifest from GitHub.

        ```

>         curl https://raw.githubusercontent.com/IBM/ibm-block-csi-operator/v1.10.0/deploy/installer/generated/host-definition.yaml > host-definition.yaml

Reviewer: @matancarmeli7 could you point me to the PR where this path exists?

Author: He said it would be in the same place as the installer so I put it in like that for now. i also assumed the file name

Author: updated

        ```

>     2. Verify that the hostdefinition is in the _Ready_ phase.

Reviewer: @matancarmeli7 how could this be ready if no yaml was applied?

Author: changed

Reviewer: I still don't see any "apply" (like e.g. [step 2.c.](https://www.ibm.com/docs/en/stg-block-csi-driver/1.9.0?topic=driver-installing-github))

Author: added

Reviewer: thanks, please write here once it's available for review

Author: available

        ```
        $> kubectl get hostdefinition
        NAME                  AGE    PHASE   STORAGE          HOST

>         <k8s_resource_name1>  102m   Ready   <storage_name>   <host_name_w1>

Author: updated

>         <k8s_resource_name2>  102m   Ready   <storage_name>   <host_name_w2>

Reviewer: I don't think `w2` is clear - how about simply `host_name2`?

Reviewer 2: +1

Author: updated

>         <k8s_resource_name2>  102m   Ready   <storage_name>   <host_name_w2>

Reviewer: is `storage_name` really the name of the storage system? or just its management address?

Reviewer 2: the `management address`

Reviewer: then let's use a placeholder accordingly

Reviewer 2: you are talking about the `Storage` field?

Reviewer: the example placeholders of it, yes

Author: <storage_address> ?
<storage_management_address>?

Author: updated

Reviewer: I think <[management_address](https://github.com/IBM/ibm-block-csi-driver/blob/v1.9.0/deploy/kubernetes/examples/demo-secret.yaml#L8)> is most accurate

Author: updated

        ```

        If in an `Error` state, a retry can be forced using TBD.
    

>     2.  (Optional) Update the prefix and connectivity fields in the `HostDefiner.yaml`.

Reviewer: wouldn't it be better to place this step **before** the auto-definitions are complete?

Author: changed

        |Field|Description|
        |---------|--------|
        |prefix|Adds a prefix to the hosts defined by the CSI driver.|

>         |connectivity|Selects the connectivity type for the host ports.<br>Possible input values are:<br>- `iscsi` for use with iSCSI connectivity<br>- `fc` for use with Fibre Channel over SCSI connectivity<br>- `nvme` for use with Fibre Channel over NVMe connectivity (Spectrum Virtualize storage systems only)<br>By default, this field is blank and the driver selects the strongest of available connectivity options.|

Author: yes - this was a typo - fixed.

>         |connectivity|Selects the connectivity type for the host ports.<br>Possible input values are:<br>- `iscsi` for use with iSCSI connectivity<br>- `fc` for use with Fibre Channel over SCSI connectivity<br>- `nvme` for use with Fibre Channel over NVMe connectivity (Spectrum Virtualize storage systems only)<br>By default, this field is blank and the driver selects the strongest of available connectivity options.|

Author: fixed
